core/verify_manager.go

```diff
@@ -0,0 +1,150 @@
+	close(vm.exitCh)
+}
+
+func (vm *VerifyManager) mainLoop(header *types.Header) {
+	//Start verify task from H to H-11 if need.
+	vm.NewBlockVerifyTask(header)
+	prune := time.NewTicker(time.Second)
+	defer prune.Stop()
+	for {
+		select {
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			rawdb.WriteTrustBlockHash(vm.bc.db, hash)
+			delete(vm.tasks, hash)
+		case <-prune.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > 15 {
+					delete(vm.tasks, hash)
+					close(task.terminalCh)
+				}
+			}
+		case <-vm.exitCh:
+			return
+		}
+	}
+}
+
+func (vm *VerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; i <= 11; i++ {
+		hash := header.Hash()
+		//if verified cache record that this block has been verified, skip.
+		if _, ok := vm.verifiedCache.Get(hash); ok {
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
+			continue
+		}
+		//if there already has a verify task for this block, skip.
+		if _, ok := vm.tasks[hash]; ok {
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
+			continue
+		}
+		//if verified storage record that this block has been verified, skip.
+		if rawdb.ReadTrustBlockHash(vm.bc.db, hash) {
+			vm.cacheBlockVerified(hash)
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
+			continue
+		}
+		diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+		//if this block has no diff, there is no need to verify it.
+		var err error
+		if diffLayer == nil {
+			if diffLayer, err = vm.bc.GenerateDiffLayer(hash); err != nil {
+				log.Error("failed to get diff layer", "block", hash, "number", header.Number, "error", err)
+				header = vm.bc.GetHeaderByHash(header.ParentHash)
+				continue
+			}
+		}
+		diffHash, err := GetTrustedDiffHash(diffLayer)
+		if err != nil {
+			log.Error("failed to get diff hash", "block", hash, "number", header.Number, "error", err)
+			header = vm.bc.GetHeaderByHash(header.ParentHash)
+			continue
+		}
+		verifyTask := NewVerifyTask(diffHash, header, vm.peers, vm.bc.db, vm.verifyCh, vm.allowUntrustedVerify)
+		vm.tasks[hash] = verifyTask
+		header = vm.bc.GetHeaderByHash(header.ParentHash)
+	}
+}
+
+func (vm *VerifyManager) cacheBlockVerified(hash common.Hash) {
+	if vm.verifiedCache.Len() >= verifiedCacheSize {
+		vm.verifiedCache.RemoveOldest()
+	}
+	vm.verifiedCache.Add(hash, true)
+}
+
+//CheckAncestorVerified function check whether H-11 block has been verified or it's a empty block.
+//If not, the blockchain should stop to insert new block.
+func (vm *VerifyManager) CheckAncestorVerified(header *types.Header) bool {
+	//find header of H-11 block.
+	header = vm.bc.GetHeaderByNumber(header.Number.Uint64() - 11)
+	//If start from genesis block, there has not a H-11 block.
+	if header == nil {
+		return true
+	}
+	hash := header.Hash()
+	//check whether H-11 block is a empty block.
+	parent := vm.bc.GetHeaderByHash(hash)
+	if header.TxHash == (common.Hash{}) && header.Root == parent.Root {
```

Comment:
it is not `header.TxHash == (common.Hash{})`, it is empty hash, but not `common.Hash{}`